core/slice.go

```diff
@@ -306,80 +318,111 @@ func (sl *Slice) Append(header *types.WorkObject, domTerminus common.Hash, domOr
+				}
 
-			conversionAmountInQuai := big.NewInt(0)
-			for _, etx := range newInboundEtxs {
-				// If the etx is conversion
-				if types.IsConversionTx(etx) {
-					value := etx.Value()
-					// If to is in Qi, convert the value into Qi
-					if etx.To().IsInQiLedgerScope() {
-						conversionAmountInQuai = new(big.Int).Add(conversionAmountInQuai, value)
-					}
-					// If To is in Quai, convert the value into Quai
-					if etx.To().IsInQuaiLedgerScope() {
-						value = misc.QiToQuai(parent, parentExchangeRate, parent.MinerDifficulty(), value)
-						conversionAmountInQuai = new(big.Int).Add(conversionAmountInQuai, value)
-					}
+				if header.ExchangeRate().Cmp(prevBlock.ExchangeRate()) > 0 {
+					exchangeRateIncreasing = true
 				}
 			}
 
-			// compute and write the conversion flow amount based on the current block
-			currentBlockConversionFlowAmount := sl.hc.ComputeConversionFlowAmount(parent, new(big.Int).Set(conversionAmountInQuai))
-			if block.ConversionFlowAmount().Cmp(currentBlockConversionFlowAmount) != 0 {
-				return nil, fmt.Errorf("invalid conversion flow amount used (remote: %d local: %d)", block.ConversionFlowAmount(), conversionFlowAmount)
-			}
+			// Compute the cubic flow discount
+			discountedConversionAmount := misc.ApplyCubicDiscount(conversionAmountInQuai, header.ConversionFlowAmount())
+			discountedConversionAmountInInt, _ := discountedConversionAmount.Int(nil)
 
-			///////// Step 2 /////////
 			// Apply K Quai discount before applying the quadratic discount, conversionAmount = (1-kQuaiDiscount)*conversionAmountInHash
-			kQuaiDiscount := parent.KQuaiDiscount()
-			conversionAmountAfterKQuaiDiscount := new(big.Int).Mul(conversionAmountInQuai, new(big.Int).Sub(big.NewInt(100), kQuaiDiscount))
+			kQuaiDiscount := header.KQuaiDiscount()
+			conversionAmountAfterKQuaiDiscount := new(big.Int).Mul(discountedConversionAmountInInt, new(big.Int).Sub(big.NewInt(params.KQuaiDiscountMultiplier), kQuaiDiscount))
+			conversionAmountAfterKQuaiDiscount = new(big.Int).Div(conversionAmountAfterKQuaiDiscount, big.NewInt(params.KQuaiDiscountMultiplier))
 
-			originalEtxValues := make([]*big.Int, len(newInboundEtxs))
+			// stores the original etx values before the conversion so that it can be used once the
+			// new exchange rate is calculated
+			etxValuesBeforeConversion := make([]*big.Int, len(newInboundEtxs))
+
+			realizedConversionAmountInQuai := big.NewInt(0)
 
 			for i, etx := range newInboundEtxs {
 
-				// store the original etx values
-				originalEtxValues[i] = new(big.Int).Set(etx.Value())
+				originalValue := new(big.Int).Set(etx.Value())
 
 				// If the etx is conversion
 				if types.IsConversionTx(etx) {
-					value := etx.Value()
+					value := new(big.Int).Set(etx.Value())
+
+					// Apply the conversion flow discount to all transactions
+					value = new(big.Int).Mul(value, discountedConversionAmountInInt)
+					value = new(big.Int).Div(value, conversionAmountInQuai)
```

Comment:
Here you would have divide by 0 if it's possible for the ETX to convert 0 value. We just need to make sure that it's not possible

Reply:
we check for minimum